docs/architecture-decisions/2023-01-12_smart-contract-deployment.md

```diff
@@ -0,0 +1,375 @@
+- [Other patterns](https://ethereum.org/en/developers/docs/smart-contracts/upgrading/#what-is-a-smart-contract-upgrade).
+
+Resolving off-chain can be done through a variety of ways, e.g.:
+
+- Create a contract discovery service API that allows new contracts to be registered by API call after being deployed.
+- If the contract creator account is always known (i.e. ephemeral creator accounts aren't being used) then it's possible to identify a contract by name by encoding a payload into the application creation transaction note and then using the indexer to find the application creation transactions for a given account and working backwards to find the relevant application ID.
+- If the contract creator account is always known (i.e. ephemeral creator accounts aren't being used) then it's possible to identify a contract by name by encoding the name in to one of the [application params](https://developer.algorand.org/docs/rest-apis/algod/v2/#applicationparams) (e.g. approval program or using a global state variable) and looking up the `created-apps` property when [retrieving an account via algod](https://developer.algorand.org/docs/rest-apis/algod/v2/#account).
+- Create a contract discovery service API that scans the blockchain transactions ([example](https://developer.algorand.org/articles/developer-preview-of-conduit-a-new-way-to-access-algorand-chain-data/)) to automatically maintain a record of smart contract IDs.
+- Use CI/CD variables ([example](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter)) to propagate the deployed app ID to the configuration on the deployment of the client (assuming the same pipeline is used for deployment of smart contract and client(s)).
+- Manually set the application ID via the CI/CD pipeline of the clients with environment-specific values (that are passed in via environment variables) and change them to reflect any changes to the smart contracts (note: this isn't sustainable for frequently changing contracts, but is a simple method for infrequently changing contracts).
+
+In order to provide the simplest implementation possible to handle this complexity, AlgoKit v1 will:
+
+- Add payloads to application creation transactions to allow them to be resolved by name for a known creator account using the indexer.
+- Allow application IDs to be resolved by name using environment variables for situations where a developer is using ephemeral creator accounts.
+- On-chain resolution will be left for developers to implement (likely by hardcoding or Global-state-based configurable application IDs).
+
+More sophisticated options could potentially by implemented in the future.
+
+### Automated vs manual deployments
+
+In order to support Continuous Delivery for smart contracts there is a need to support automated deployments of smart contracts via deployment pipeline.
```

Comment:
how important is this for algokit v1? Are we talking about CD while developing against Testnet or something? I am imagining that there are few use cases where people would actually have upgradeable smart contracts that they want to use CD to upgrade. I _can_ imagine such a case but I don't think it's a common one.

Reply:
I do of course see value in guiding a healthy deployment against Mainnet, I'm just not sure about the CD part.

Reply:
One argument was that it is best if the same process is used to deploy on TestNet vs deploy on MainNet, so that there are fewer risks when moving into production.

Reply:
I don't think it should be a CI/CD pipeline necessarily for MainNet, but more a support that allow easily deployment as when done locally on TestNet/Sandbox.
Also, I think this should not allow for MainNet upgrade of smart contract, because MainNet upgrade of smart contract should only be done with secure keys and algokit does not have a secure-enough way of storing keys for that purpose.

Reply:
> One argument was that it is best if the same process is used to deploy on TestNet vs deploy on MainNet, so that there are fewer risks when moving into production.

💯 

Exactly Fabrice, it's about providing confidence and rigour by ensuring that only code that has been through a defined process that incrementally increases confidence and reduces risk can be deployed to production (MainNet).

Is it essential for v1? Probably not, it's something I'm passionate about though :)